### PR TITLE
Fixes broken LibGit2.rebase!

### PR DESCRIPTION
**Pull request by simonbyrne**
`LibGit2.rebase!` is broken, causing problems with `Pkg.update()`
after `PkgDev.tag()`, see
https://discourse.julialang.org/t/how-to-properly-add-releases-with-pkgdev/961/15?u=simonbyrne

There were two problems:
- `upstream` was used as a variable name and a function call
- if the patch was redundant (`EAPPLIED` return code), `rebase!` would
throw an error.

### CODE REVIEW

File: base/libgit2/rebase.jl [hunk 1]
```diff
@@ -43,9 +43,12 @@ end
 
 function commit(rb::GitRebase, sig::GitSignature)
     oid_ptr = Ref(Oid())
-    @check ccall((:git_rebase_commit, :libgit2), Cint,
+    err = ccall((:git_rebase_commit, :libgit2), Error.Code,
                   (Ptr{Oid}, Ptr{Void}, Ptr{SignatureStruct}, Ptr{SignatureStruct}, Ptr{UInt8}, Ptr{UInt8}),
-                   oid_ptr, rb.ptr, C_NULL, sig.ptr, C_NULL, C_NULL)
+                oid_ptr, rb.ptr, C_NULL, sig.ptr, C_NULL, C_NULL)
```
Comment (tkelman) [inline base/libgit2/rebase.jl]: alignment probably shouldn't change here?

**Comment by tkelman** [comment]
yay, Simon to the rescue. can we construct this situation locally and test for it in test/libgit2.jl or test/pkg.jl ?

**Comment by simonbyrne** [comment]
Yes, we probably should do it, and I imagine it should be possible, but I'm not sure how best to do it. Any suggestions?

**Comment by tkelman** [comment]
create 2 local test repos with the same content but via different series of commits so that merging from an the on-disk remote that points to the other repo wouldn't be a fast forward but would trigger this bug?

**Comment by simonbyrne** [comment]
I'm still not sure why this went undiscovered for so long: it should have thrown a more informative error.

**Comment by tkelman** [comment]
I'm not surprised, the libgit2 code is full of this kind of thing, it could use a pretty thorough review, refactoring, and docs.

**Comment by simonbyrne** [comment]
I've added a test: it fails earlier and now passes.

File: base/libgit2/rebase.jl [hunk 2]
```diff
@@ -41,11 +41,23 @@ function Base.next(rb::GitRebase)
...
     oid_ptr = Ref(Oid())
-    @check ccall((:git_rebase_commit, :libgit2), Cint,
-                  (Ptr{Oid}, Ptr{Void}, Ptr{SignatureStruct}, Ptr{SignatureStruct}, Ptr{UInt8}, Ptr{UInt8}),
-                   oid_ptr, rb.ptr, C_NULL, sig.ptr, C_NULL, C_NULL)
+    try
+        @check ccall((:git_rebase_commit, :libgit2), Error.Code,
+                     (Ptr{Oid}, Ptr{Void}, Ptr{SignatureStruct}, Ptr{SignatureStruct}, Ptr{UInt8}, Ptr{UInt8}),
+                      oid_ptr, rb.ptr, C_NULL, sig.ptr, C_NULL, C_NULL)
+    catch err
+        err.code == Error.EAPPLIED && return nothing
```
Comment (tkelman) [inline base/libgit2/rebase.jl]: is returning nothing acceptable for callers? would returning the head be any more appropriate?

**Comment by tkelman** [comment]
cool. that test passes without internet access, right? looks like it should, but if not then it should be moved to libgit2-online.

**Comment by simonbyrne** [comment]
Yes, it just tests rebase locally (as it is a local operation).

File: base/libgit2/rebase.jl (hunk 2 above)
Comment (simonbyrne) [inline base/libgit2/rebase.jl]: probably, but I can't figure out how to do that.

Comment (tkelman) [inline base/libgit2/rebase.jl]: hm okay. leave a todo comment maybe?

**Review by tkelman — APPROVED** [review]

**Pull request merged**